File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        vector<vector<int>> buckets(nums.size() + 1);
        map<int, int> freq;
        for (auto num: nums) freq[num]++;
        for (auto num: freq) buckets[num.second].push_back(num.first);

        vector<int> res;
        for (int i = nums.size(); i >= 0; i--) {
            for (auto num: buckets[i]) {
                res.push_back(num);
                k--;
                if (k == 0) return res;
            }
        }
        return res;
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        
        cnts = Counter(nums)
        minHeap = []
        for key,val in cnts.items():
            if len(minHeap) < k:
                heappush(minHeap, (val, key))
            else:
                if minHeap[0][0] < val:
                    heappop(minHeap)
                    heappush(minHeap, (val, key))
        
        output = []
        while minHeap:
            output.append(minHeap[0][1])
            heappop(minHeap)
        
        return output
                

